File: python_feed_triplex/common/path.py
"""Path resolution helpers."""

from __future__ import annotations

from pathlib import Path

from python_feed_triplex.common import constants


def resolve_path(dir_name: str, file_name: str) -> Path:
    working_dir = Path.cwd()
    executable_dir = Path(__file__).resolve().parents[2]
    candidates = [
        working_dir / dir_name / file_name,
        working_dir / file_name,
        executable_dir / file_name,
        executable_dir / dir_name / file_name,
        executable_dir.parent / dir_name / file_name,
    ]
    for candidate in candidates:
        if candidate.exists():
            return candidate
    raise FileNotFoundError(f"Unable to locate {file_name}")


def resolve_config_path() -> Path:
    return resolve_path(constants.CONFIG_DIR_NAME, constants.CONFIG_FILE_NAME)


def resolve_logger_config_path() -> Path:
    return resolve_path(constants.LOGGER_CONFIG_DIR_NAME, constants.LOGGER_CONFIG_FILE_NAME)


def resolve_app_root() -> Path:
    config_path = resolve_config_path()
    config_dir = config_path.parent
    if config_dir.name == constants.CONFIG_DIR_NAME:
        return config_dir.parent
    return config_dir

File: real_engine/common/path.go
package common

import (
	"fmt"
	"os"
	"path/filepath"
)

func ResolveConfigPath() (string, error) {
	return resolvePath(ConfigDirName, ConfigFileName)
}

func ResolveLoggerConfigPath() (string, error) {
	return resolvePath(LoggerConfigDirName, LoggerConfigFileName)
}

func ResolveAppRoot() (string, error) {
	configPath, err := ResolveConfigPath()
	if err != nil {
		return "", err
	}

	configDir := filepath.Dir(configPath)
	if filepath.Base(configDir) == ConfigDirName {
		return filepath.Dir(configDir), nil
	}
	return configDir, nil
}

func resolvePath(dirName, fileName string) (string, error) {
	workingDir, _ := os.Getwd()
	executablePath, _ := os.Executable()
	executableDir := filepath.Dir(executablePath)

	candidates := []string{
		filepath.Join(workingDir, dirName, fileName),
		filepath.Join(workingDir, fileName),
		filepath.Join(executableDir, fileName),
		filepath.Join(executableDir, dirName, fileName),
		filepath.Join(executableDir, "..", dirName, fileName),
	}

	for _, candidate := range candidates {
		if _, err := os.Stat(candidate); err == nil {
			return candidate, nil
		}
	}

	return "", fmt.Errorf("unable to locate %s", fileName)
}
